File: algorithms/tree/search/bfs.py
# Breadth First Search
from collections import deque
from algorithms.tree.design.tree import TreeNode

# Time Complexity:      O(n)
# Space Complexity :    O(n)
def bfs(root):
    if not root:
        return []

    queue = deque([root])
    result = []

    while queue:
        node = queue.popleft()
        result.append(node.val)

        if node.left:
            queue.append(node.left)
        if node.right:
            queue.append(node.right)

    return result


# Tests
root = TreeNode(1)
root.left = TreeNode(2)
root.right = TreeNode(3)

root.left.left = TreeNode(4)
root.left.right = TreeNode(5)

root.right.right = TreeNode(6)

print(bfs(root))

File: cpp_algorithms/tree/search/bfs.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
#define pb push_back

// Binary Tree Node
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;

    TreeNode(int x) {
        val = x;
        left = nullptr;
        right = nullptr;
    }
};

// Breadth First Search
//
// Time Complexity:  O(n)
// Space Complexity: O(n)

vector<int> bfs(TreeNode* root) {
    if (root == nullptr) {
        return {};
    }

    queue<TreeNode*> q;
    q.push(root);

    vector<int> result;

    while (!q.empty()) {
        TreeNode* node = q.front();
        q.pop();

        result.push_back(node->val);

        if (node->left != nullptr) {
            q.push(node->left);
        }

        if (node->right != nullptr) {
            q.push(node->right);
        }
    }

    return result;
}

int main() {
    TreeNode* root = new TreeNode(1);

    root->left = new TreeNode(2);
    root->right = new TreeNode(3);

    root->left->left = new TreeNode(4);
    root->left->right = new TreeNode(5);

    root->right->right = new TreeNode(6);

    vector<int> result = bfs(root);

    for (int x : result) {
        cout << x << ' ';
    }

    cout << '\n';

    return 0;
}
